File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* list1 = l1;
        ListNode* list2 = l2;
        ListNode* resHead=NULL;
        ListNode* res=NULL;
        int sum=0, carry=0, add=0;
        while(list1 != NULL && list2 != NULL) {
            sum = list1->val + list2->val + carry;
            add = sum % 10;
            carry = sum / 10;
            ListNode* node = new ListNode(add);
            if (resHead == NULL) {
                resHead = node;
                res = resHead;
            } else {
                res->next = node;
                res = res->next;
            }
            list1 = list1->next;
            list2 = list2->next;
        }
        while(list1 != NULL) {
            sum = list1->val + carry;
            add = sum % 10;
            carry = sum / 10;
            ListNode* node = new ListNode(add);
            if (resHead == NULL) {
                resHead = node;
                res = resHead;
            } else {
                res->next = node;
                res = res->next;
            }
            list1 = list1->next;
        }
        while(list2 != NULL) {
            sum = list2->val + carry;
            add = sum % 10;
            carry = sum / 10;
            ListNode* node = new ListNode(add);
            if (resHead == NULL) {
                resHead = node;
                res = resHead;
            } else {
                res->next = node;
                res = res->next;
            }
            list2 = list2->next;
        }
        if (carry > 0) {
            ListNode* node = new ListNode(carry);
            res->next = node;
            res = res->next;
        }
        return resHead;
    }
};

File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        l3 = ListNode(0)
        head = l3
        carry = 0

        while l1 and l2:
            val = l1.val + l2.val + carry
            carry = int(val / 10)
            val = int(val % 10)
            

            node = ListNode(val)
            l3.next = node
            l3 = l3.next

            l1 = l1.next
            l2 = l2.next
        
        while l1:
            val = l1.val + carry
            carry = int(val / 10)
            val = int(val % 10)
            

            node = ListNode(val)
            l3.next = node
            l3 = l3.next

            l1 = l1.next
        
        while l2:
            val = l2.val + carry
            carry = int(val / 10)
            val = int(val % 10)
            

            node = ListNode(val)
            l3.next = node
            l3 = l3.next

            l2 = l2.next
        
        if carry:
            node = ListNode(carry)
            l3.next = node
            l3 = l3.next
        
        return head.next

        
